style: responsive webui [DET-4417, DET-4420]

hkang1 opened the pull request:
## Description

Update the WebUI to be mobile friendly. Key areas of updates:

- Primary navigation
- Support for Responsive tables
- Filter groups in Section headers

<img width="362" alt="Screen Shot 2020-10-28 at 11 49 00 AM" src="https://user-images.githubusercontent.com/220971/97476450-05aa0180-1914-11eb-9da8-0ea3c9a006b8.png">
<img width="362" alt="Screen Shot 2020-10-28 at 11 49 11 AM" src="https://user-images.githubusercontent.com/220971/97476456-0773c500-1914-11eb-8e2d-f4be989064f2.png">
<img width="362" alt="Screen Shot 2020-10-28 at 11 50 28 AM" src="https://user-images.githubusercontent.com/220971/97476460-08a4f200-1914-11eb-8b30-e0209e4a8356.png">
<img width="362" alt="Screen Shot 2020-10-28 at 11 50 52 AM" src="https://user-images.githubusercontent.com/220971/97477314-160eac00-1915-11eb-9e2f-0f7b12824dd5.png">
<img width="362" alt="Screen Shot 2020-10-28 at 11 50 52 AM" src="https://user-images.githubusercontent.com/220971/97476485-0fcc0000-1914-11eb-8f44-8d50d53012e8.png">
<img width="362" alt="Screen Shot 2020-10-28 at 11 51 11 AM" src="https://user-images.githubusercontent.com/220971/97476489-10fd2d00-1914-11eb-87db-d26167003351.png">


<!---
Lead with the intended commit body in this description field. For breaking changes,
please include "BREAKING CHANGE:" at the beginning of your commit body.
At minimum, this section should include a bracketed reference to the Jira ticket,
e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.
-->


## Test Plan

- Check every WebUI page to ensure they render properly in desktop, tablet and mobile views (check to make sure they are fluid as well)
  - Login page
  - Dashboard
  - Experiment List
  - Experiment Detail
  - Trial Detail
  - Trial Logs
  - Task List
  - Cluster
  - Master Logs
- Check for the tables to render properly for mobile views.
- Check for the filter groups to render properly. (the filter options that appear across from the Section headers. For example the `User` or `State` filters on the experiment list page)

<!---
Describe the situations in which you've tested your change, and/or a screenshot as appropriate.
Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.
-->


## Commentary (optional)

<!---
Use this section of your description to add context to the PR. Could be for particularly
tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->


## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.


<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

hkang1 commented on webui/react/src/hooks/useNotebookLauncher.ts:
```diff
@@ -0,0 +1,39 @@
+import { useCallback } from 'react';
```
convert to module (nothing in here is React specific)

hkang1 commented on webui/react/.storybook/main.js:
```diff
@@ -19,7 +19,7 @@ module.exports = {
     '@storybook/addon-knobs',
   ],
-  stories: ['../src/**/*.stories.(ts|tsx)'],
+  stories: ['../src/**/*.stories.@(ts|tsx)'],
```
breaking change for Storybook 6

hamidzr reviewed (commented):
Great job with making the UI responsible. This should make the WebUI a lot more usable on phones and smaller windows. I added all the comments  I had. I realized you had 2 or 3 comments open as well (eg turning that hook into a module). 

- does make storybook-build passes for you? it is passing in CI but runs into an issue for me

![Screen Capture_select-area_20201102144223](https://user-images.githubusercontent.com/12127420/97930048-75a6f600-1d1f-11eb-8ec6-be7343427591.png)



- no need to mention `/det` in routes (would probably need a rebase)

### style nits

- I think the avatar could be a bit bigger compared to the logo
![Screen Capture_select-area_20201102151734](https://user-images.githubusercontent.com/12127420/97929702-9458bd00-1d1e-11eb-87c2-9d85ac0c4866.png)
alignment here
![Screen Capture_select-area_20201102151919](https://user-images.githubusercontent.com/12127420/97929910-1c3ec700-1d1f-11eb-88e0-67dde3c4eb91.png)
the page itself is horizontally scrollable for experiments list
![Screen Capture_select-area_20201102152049](https://user-images.githubusercontent.com/12127420/97929921-24970200-1d1f-11eb-9722-48937401cf28.png)

hkang1 commented on webui/react/config-overrides.js:
```diff
@@ -13,6 +14,8 @@ const path = require('path');
 const webpack = require('webpack');
 const jestConfig = require('./jest.config');
 
+const fs = require('fs');
```
removed

hkang1 commented on webui/react/src/pages/SignIn.tsx:
```diff
@@ -73,13 +74,16 @@ const SignIn: React.FC = () => {
-        <Logo type={LogoTypes.OnLightVertical} />
-        <DeterminedAuth />
+  return auth.checked ? (
+    <Page>
```
yeah probably, some merge conflict came up here so it's resolved now

hkang1 commented on webui/react/src/components/NavigationTabbar.stories.tsx:
```diff
@@ -0,0 +1,42 @@
+};
+
+export const Default = (): React.ReactNode => (
+  <div style={{
+    border: 'solid 1px #cccccc',
+    display: 'flex',
+    flexDirection: 'column',
+    height: 480,
+    position: 'relative',
+    width: 320,
+  }}>
+    <div style={{
+      alignItems: 'center',
+      display: 'flex',
+      flexGrow: 1,
+      justifyContent: 'center',
+    }}>Content Area</div>
```
using a storybook add-on to introduce responsive view into our stories

hkang1 commented:
> Great job with making the UI responsible. This should make the WebUI a lot more usable on phones and smaller windows. I added all the comments I had. I realized you had 2 or 3 comments open as well (eg turning that hook into a module).
> 
> * does make storybook-build passes for you? it is passing in CI but runs into an issue for me
> 
> ![Screen Capture_select-area_20201102144223](https://user-images.githubusercontent.com/12127420/97930048-75a6f600-1d1f-11eb-8ec6-be7343427591.png)
> 
This worked for me when I tried it just now. Based on the error, it looks like the config-override of bringing global SASS of `@import global` was not coming into play for storybook? Try again?

> * no need to mention `/det` in routes (would probably need a rebase)
> 
Yep updated them now that your changes are in. Thanks for calling it out!

> ### style nits
> * I think the avatar could be a bit bigger compared to the logo
>
>   ![Screen Capture_select-area_20201102151734](https://user-images.githubusercontent.com/12127420/97929702-9458bd00-1d1e-11eb-87c2-9d85ac0c4866.png)
>
>   alignment here
>   ![Screen Capture_select-area_20201102151919](https://user-images.githubusercontent.com/12127420/97929910-
1c3ec700-1d1f-11eb-88e0-67dde3c4eb91.png)
>
I will do another round of style refresh as another task for these style corrections, just because there are a few other places that should get addressed (Page, Page Header, Section Options, etc).


>   the page itself is horizontally scrollable for experiments list
>   ![Screen Capture_select-area_20201102152049](https://user-images.githubusercontent.com/12127420/97929921-24970200-1d1f-11eb-9722-48937401cf28.png)

Ah good find, the TableBatch was the culprit! Fixed.

hamidzr reviewed (approved):
Nice work. LGTM minus one blocking comment about `external` flag.

- suggestion: pass `Link` props up as part of actions in `ActionSheet`

hkang1 commented on webui/react/src/hooks/useResize.ts:
```diff
@@ -16,8 +16,8 @@ const defaultResizeInfo = {
 
-const useResize = (ref: RefObject<HTMLElement>): ResizeInfo => {
-  const element = ref.current;
+const useResize = (ref?: RefObject<HTMLElement> | undefined): ResizeInfo => {
```
lol, yep 👍

hkang1 commented on webui/react/src/components/NavigationTabbar.tsx:
```diff
@@ -0,0 +1,87 @@
+            onClick: () => handleLaunchNotebook(true),
+          },
+          { icon: 'logs', label: 'Master Logs', path: '/logs', popout: true },
+          { icon: 'docs', label: 'Docs', path: '/docs', popout: true },
+          { icon: 'cloud', label: 'API (Beta)', path: '/docs/rest-api/', popout: true },
```
👍

hkang1 commented on webui/react/src/components/IconFilterButtons.tsx:
```diff
@@ -25,21 +25,25 @@ const FilterButtons: React.FC<Props> = ({ buttons, onClick }: Props) => {
-        );
-      })}
+      <div>
+        <div className={css.buttons}>
```
good catch, extra useless div in there 🤷

hkang1 commented on webui/react/src/components/ResponsiveTable.tsx:
```diff
@@ -0,0 +1,33 @@
+import useResize from 'hooks/useResize';
+
+/* eslint-disable-next-line @typescript-eslint/ban-types */
+type ResponsiveTable = <T extends object>(props: TableProps<T>) => JSX.Element;
```
When used in a situation like:
```
ResponsiveTable<TrialItem>
```
`TrialItem` doesn't conform to `Record<string, unknown>`

hkang1 commented on webui/react/src/utils/task.ts:
```diff
@@ -1,9 +1,32 @@
 import { terminalCommandStates } from 'utils/types';
+import { commandToTask } from 'utils/types';
+
+export const launchNotebook = async (slots: number): Promise<void> => {
```
I'll put it in `/services/utils`.

hkang1 replied:
actually, discovered that this is a high level function which touches a lot of different areas. Moving it into services (no matter what file) creates a circular dependency.
